Allow forward functions with single output to return Variable

malvika2147 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23803 Allow forward functions with single output to return Variable**

Summary:
Custom `forward()` can return a `Variable` in case of single outputs instead of returning a `variable_list` of size 1. 

Test Plan: Modified tests involving single output forward functions.

Differential Revision: [D16673857](https://our.internmc.facebook.com/intern/diff/D16673857)

Commit:
Allow forward functions with single output to return Variable

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

ezyang commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -56,10 +56,13 @@ TORCH_API variable_list _wrap_outputs(
 // Example backward call:
 // y[0].sum().backward();
+template<typename X, typename... Args>
+using forward_t = decltype(X::forward(nullptr, std::declval<Args>()...));
```
You probably didn't mean to attach the docblock above to `forward_t`. (Speaking of which, would be nice to have a doc for `forward_t` too!)

ezyang commented:
@smessmer could you please help review the C++ template metaprogramming in this diff?

ezyang reviewed (approved):
Seems to work, but I don't know too much about how the template metaprogramming works here.

Commit:
Update on "Allow forward functions with single output to return Variable"


Summary:
Custom `forward()` can return a `Variable` in case of single outputs instead of returning a `variable_list` of size 1. 

Test Plan: Modified tests involving single output forward functions.

Commit:
Update on "Allow forward functions with single output to return Variable"


Summary:
Custom `forward()` can return a `Variable` in case of single outputs instead of returning a `variable_list` of size 1. 

Test Plan: Modified tests involving single output forward functions.

Differential Revision: [D16673857](https://our.internmc.facebook.com/intern/diff/D16673857)

smessmer commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -182,7 +198,8 @@ variable_list Function<T>::apply(Args&&... args) {
   }
 
-  variable_list outputs;
+  typedef forward_t<X, Args...> forward_return_t;
```
nit: `using` is more readable than `typedef`

smessmer commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -159,9 +169,15 @@ template <typename... Args>
 }
 
+template <typename T>
+typename std::enable_if<std::is_same<T, variable_list>::value, T&>::type to_output_type(variable_list& output_list) { return output_list; }
```
nit: we have `guts::enable_if_t` to avoid the `typename`

smessmer reviewed (commented):
metaprogramming looks good

Commit:
Update on "Allow forward functions with single output to return Variable"


Summary:
Custom `forward()` can return a `Variable` in case of single outputs instead of returning a `variable_list` of size 1. 

Test Plan: Modified tests involving single output forward functions.

Differential Revision: [D16673857](https://our.internmc.facebook.com/intern/diff/D16673857)

Commit:
Update on "Allow forward functions with single output to return Variable"


Summary:
Custom `forward()` can return a `Variable` in case of single outputs instead of returning a `variable_list` of size 1. 

Test Plan: Modified tests involving single output forward functions.

Differential Revision: [D16673857](https://our.internmc.facebook.com/intern/diff/D16673857)
